src/pipx/shared_libs.py

```diff
@@ -30,6 +31,15 @@ def __init__(self) -> None:
         # i.e. python_path is ~/.local/pipx/shared/python
         self._site_packages: Optional[Path] = None
         self.has_been_updated_this_run = False
+        # TODO: remove setuptools (wheel?)
+        self.required_packages = [
+            "pip",
+            "wheel",
+            "packaging",
+            "importlib-metadata",
+            "setuptools",
```

Comment:
I just realized there are no version constraints on these package specifiers. I have been getting bit on some semver changes on my other projects so it has been on my mind. However, thinking about these a little, it's likely we always want to be on the latest version of all of them since they all work in tandem (i.e. an update to the wheel format will be reflected across.

> Adding more libraries to the shared libs does increase the chance that we will conflict with requirements in the venvs.

I wonder what would happen here. The shared lib packages are added via the pth file, and aren't part of the requirements of the package being installed. I assume pip (and its new dependency resolver) only applies to the requirements/package being installed (and its dependencies). The shared lib pip invocation is separate from app package installation and manipulation, so maybe there won't be any issues here?

Reply:
> I wonder what would happen here. The shared lib packages are added via the pth file, and aren't part of the requirements of the package being installed. I assume pip (and its new dependency resolver) only applies to the requirements/package being installed (and its dependencies). The shared lib pip invocation is separate from app package installation and manipulation, so maybe there won't be any issues here?

It depends on how and when the shared libs are injected. pip determines whether a package is installed by inspecting directories on `sys.path`, so if the shared libs are injected when `pip install` is run, pip will consider those packages as “installed”. If you’re going to add version constraints to those libraries, they could end up conflicting with applications users want to install.

I think maybe the best way to implement `venv_metadata_inspecor.py` is to run it as a [zipapp](https://docs.python.org/3/library/zipapp.html) instead. pipx can create a zipapp `venv_metadata_inspecor.pyz` that contains the script and all its dependencies in a cache directory, and run `python venv_metadata_inspecor.pyz` instead. The same technique can actually be used for `pip install` to avoid potential edge cases where an application depends on a certain pip and setuptools version not compatible with pipx (not likely to happen though).

Reply:
The zipapp idea is very interesting.  Wouldn't pipx still need to be running inside of the target venv anyway for `importlib[.-]metadata.Distribution` to get information?  And wouldn't that end up with the same version conflicts?

I don't see any way to point `importlib[.-]metadata` to examine another venv than the current one, in which case I'm worried even a zipapp would encounter the same version conflicts.

Is there a way around these problems?

Reply:
Oooh once I actually wrote out "any way to point importlib[.-]metadata to examine another venv than the current one" I realized that is the key, and I found this pointer, which may show the way to do that:
https://docs.python.org/3/library/importlib.metadata.html#extending-the-search-algorithm

I'm wondering if we could hook into that to facilitate running the code in one venv and inspecting another?

Reply:
If we could do THAT, then we could make venv inspection just another part of pipx and not a separate runnable file.  That would really be cool.

Reply:
Actually, maybe all we need to do is manipulate `sys.path`.  It might be as simple as that, we just need to put the directories of the venv to be inspected in `sys.path` before asking `importlib.metadata` about distribution information.  I've been experimenting and that seems to work.

Reply:
Since I'm stuck at home alone on Christmas Eve 😆 , I've been trying the `sys.path` manipulation trick, and it works so well.  I may have a new PR to audition for you all soon.

Reply:
pipx itself does not need to be run inside the target environment, only the `venv_metadata_inspecor` script.

> Actually, maybe all we need to do is manipulate `sys.path`.

This is precisely what a zipapp does; basically `sys.path.insert(0, path_to_zip)`. But by using a zipapp, you can separate pip + setuptools and other stuff, so those other stuff would not be present in the environment when `pip install` is run, thus not resulting vesion conflicts.

I’m not quite sure how `sys.path` manipulation would work without overhauling how shared libs are managed. Currently all shared libs are installed in one virtual environment, and injecting any one of them means injecting all of them. So when you inject pip to run `pip install`, pip will still see the shared packaging etc. libs and potentially emit version conflicts. To prevent that from happening, you’ll need to install packaging etc. into a separate environment, which is basically manually re-implement zipapp.

Reply:
What I've been doing in another branch, is to have the shared libs be only what we currently have: `pip`, `setuptools`, and `wheel`.  Just to install new packages, not to inspect venvs.  Then I've made the equivalent of `venv_metadata_inspector.py` just another submodule of pipx.  The only thing I need to do that's "special" is: before cataloging `Distributions`, I swap out pipx's `sys.path` list of directories with the `sys.path` for the venv.  Then once I've cataloged all the `Distributions`, I swap pipx's `sys.path` back.

It seems to work great (I'm not importing anything during the temporary change in `sys.path`) and as a bonus, the venv inspection code is a first-class citizen of pipx, and we don't have to use subprocesses, pass around json, or use python 3.5 syntax for type-checking.

Reply:
Basically I'm only changing `sys.path` in order to change what `importlib[.-]metadata` is looking at when we get info about the distributions.  Otherwise, the code is still part of the pipx package and not in a separate venv.

Reply:
Ah OK, so you do all the inspection outside of the target virtual environment? That should work.

I believe you can also avoid `sys.path` manipulation by passing a path context, e.g. `importlib_metadata.distributions(path=path_to_venv_site_packages)`.

Reply:
> I believe you can even avoid `sys.path` manipulation altogether by passing a path context, e.g. `importlib_metadata.distributions(path=path_to_venv_site_packages)` (not sure if this would work, @jaraco is the expert on this).

That would be even better.  Let me see what I can do.

Reply:
The `path` parameter is meant to be a sequence like `sys.path` so something like `importlib.metadata.distributions(path=[path_to_venv_site_packages])` may be what you're after.

Reply:
Thanks, @jaraco.  I think I've got that working in another branch.